clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetFetcherUtils.java

```diff
@@ -0,0 +1,253 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.clients.consumer.internals;
+
+import org.apache.kafka.clients.ApiVersions;
+import org.apache.kafka.clients.consumer.OffsetResetStrategy;
+import org.apache.kafka.common.Node;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.errors.TopicAuthorizationException;
+import org.apache.kafka.common.message.ListOffsetsResponseData;
+import org.apache.kafka.common.protocol.Errors;
+import org.apache.kafka.common.requests.ListOffsetsRequest;
+import org.apache.kafka.common.requests.ListOffsetsResponse;
+import org.apache.kafka.common.utils.LogContext;
+import org.apache.kafka.common.utils.Time;
+import org.slf4j.Logger;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * Utility functions for fetching offsets, validating and resetting positions.
+ */
+class OffsetFetcherUtils {
+    private final ConsumerMetadata metadata;
+    private final SubscriptionState subscriptionState;
+    private final Time time;
+    private final ApiVersions apiVersions;
+    private final Logger log;
+
+    private final AtomicReference<RuntimeException> cachedOffsetForLeaderException = new AtomicReference<>();
+    private final AtomicInteger metadataUpdateVersion = new AtomicInteger(-1);
+
+    OffsetFetcherUtils(LogContext logContext,
+                              ConsumerMetadata metadata,
+                              SubscriptionState subscriptionState,
+                              Time time,
+                              ApiVersions apiVersions) {
+        this.log = logContext.logger(getClass());
+        this.metadata = metadata;
+        this.subscriptionState = subscriptionState;
+        this.time = time;
+        this.apiVersions = apiVersions;
+    }
+
+    /**
+     * Callback for the response of the list offset call.
+     * @param listOffsetsResponse The response from the server.
+     * @return {@link OffsetFetcherUtils.ListOffsetResult} extracted from the response, containing the fetched offsets
+     * and partitions to retry.
+     */
+    OffsetFetcherUtils.ListOffsetResult handleListOffsetResponse(ListOffsetsResponse listOffsetsResponse) {
```

Comment:
This is actually a request handler.  I might make sense to create a separated class for it (maybe nested inside one of the fetcher class), WDYT? Taking this away might make it easier to "static-ify" the class.  I've been doing that a lot for the refactored consumers.

Reply:
Definitely doable, but wonder if it's worth it in the case that we are not looking for turning this into a static class given that it maintains some state? The refactor is intentionally only extracting reusable code (not changing it), to keep the risk low. That being said, open to talk more about it, just want to make sure that any changes we introduce (other than extracting existing code) bring value that outweighs the risk